text/0249-v2-experiments.md

```diff
@@ -373,6 +379,266 @@ Disadvantages:
+npm ERR! Could not resolve dependency:
+npm ERR! peer aws-cdk-li@"^1.60.0" from my-construct-lib@1.6.0
+npm ERR! node_modules/my-construct-lib
+npm ERR!   my-construct-lib@"1.6.0" from the root project
+npm ERR!
+npm ERR! Fix the upstream dependency conflict, or retry
+npm ERR! this command with --force, or --legacy-peer-deps
+npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
+npm ERR!
+```
+
+Users can work around this by executing `npm install --force` which means that
+npm will not check for version compatibility at all, and therefore not an
+acceptable solution.
+
+##### Separate major versions
+
+This was rejected due to similar reasons as listed in the disadvantages of
+[Option 4: separate V3 that's all unstable](#option-4-separate-v3-thats-all-unstable).
+
+### Option 6: API Previews
+
+AWS CDK v2 release notes:
+
+> Starting with version 2.0.0 of the AWS CDK, all modules and members will
+> become stable. This means that from this release, we are committed to never
+> introduce breaking changes in a non-major bump.
+>
+> One of the most common feedback we hear from customers is that they love how
+> fast new features are added to the AWS CDK, we love it to. In v1, the
+> mechanism that allowed us to add new features quickly was marking them as
+> "experimental". Experimental features were not subject to semantic versioning,
+> and we allowed breaking changes to be introduced in these APIs. This is the
+> other most common feedback we hear from customers - breaking changes are not
+> ok.
+>
+> #### Introducing API Previews
+>
+> To make sure we can keep adding features fast, while keeping our commitment to
+> not release breaking changes, we are introducing a new model - API Previews.
+> APIs that we want to get in front of developers early, and are not yet
+> finalized, will be added to the AWS CDK with a specific suffix: `PreX`. APIs
+> with the preview suffix will never be removed, instead they will be deprecated
+> and replaced by either the stable version (without the suffix), or by a newer
+> preview version. For example, assume we add the method
+> `grantAwesomePowerPre1`:
+>
+> ```ts
+> /**
+>  * This methods grants awesome powers
+>  */
+> grantAwesomePowerPre1();
+> ```
+>
+> Times goes by, we get feedback that this method will actually be much better
+> if it accept a `Principal`. Since adding a required property is a breaking
+> change, we will add `grantAwesomePowerPre2()` and deprecate
+> `grantAwesomePowerPre1`:
+>
+> ```ts
+> /**
+> * This methods grants awesome powers to the given principal
+> *
+> * @param grantee The principal to grant powers to
+> */
+> grantAwesomePowerPre2(grantee: iam.IGrantable)
+>
+> /**
+> * This methods grants awesome powers
+> * @deprecated use grantAwesomePowerPre2
+> */
+> grantAwesomePowerPre1()
+> ```
+>
+> When we decide its time to graduate the API, the latest preview version will
+> be deprecated and the final version - `grantAwesomePower` will be added.
```

Comment:
Do we want to define a standard convention when we realize `grantAwesomePower` has a flaw and needs to be deprecated as well? `grantAwesomePowerPost1`? `grantAwesomePower2`?

Reply:
I don't care about the name (and suffixes are ugly), but I would like a tool-supported way to make sure that for the `vNEXT` we can re-use the "prime real estate" name for the *correct* function.

So even though in `v2`:

* `grantAwesomePower() -> @deprecated`
* `grantAwesomePowerBetter()  -> recommended`

I'd like in `v3`:

* `grantAwesomePower() ->` has the implementation of `grantAwesomePowerBetter()` from the previous version

Although maybe standardized suffixes and some compilation strategy in `jsii` are the easiest way to achieve that...